File: core/src/main/kotlin/dev/shelly/hermes/core/DiffHunkApproval.kt
package dev.shelly.hermes.core

/** Splits a unified patch into independently reviewable hunks without applying them. */
object DiffHunkApproval {
    fun expand(call: ToolCall): List<ToolCall> {
        if (call.name != "apply_patch") return listOf(call)
        val path = jsonString(call.argumentsJson, "path") ?: return listOf(call)
        val patch = jsonString(call.argumentsJson, "patch") ?: return listOf(call)
        val lines = patch.replace("\r\n", "\n").replace('\r', '\n').split('\n')
        val headers = lines.mapIndexedNotNull { index, line -> index.takeIf { line.startsWith("@@ ") } }
        if (headers.isEmpty()) return listOf(call)
        return headers.mapIndexed { index, start ->
            val end = headers.getOrNull(index + 1) ?: lines.size
            val hunk = lines.subList(start, end).joinToString("\n").trimEnd()
            ToolCall(
                id = "${call.id}:hunk-${index + 1}",
                name = "apply_patch_hunk",
                argumentsJson = "{\"path\":\"${escape(path)}\",\"hunk_index\":${index + 1}," +
                    "\"hunk_count\":${headers.size},\"hunk\":\"${escape(hunk)}\"}",
            )
        }
    }

    /** Rebuilds a full apply_patch invocation from the hunks a user approved. */
    fun collapse(approvedHunks: List<ToolCall>): ToolCall {
        require(approvedHunks.isNotEmpty()) { "At least one approved hunk is required" }
        require(approvedHunks.all { it.name == "apply_patch_hunk" }) {
            "Only apply_patch hunks can be collapsed"
        }
        val path = approvedHunks.first().let { jsonString(it.argumentsJson, "path") }
        require(!path.isNullOrBlank()) { "Approved hunk is missing a path" }
        require(approvedHunks.all { jsonString(it.argumentsJson, "path") == path }) {
            "Approved hunks must belong to one file"
        }
        val ordered = approvedHunks.sortedBy { jsonString(it.argumentsJson, "hunk_index")?.toIntOrNull() ?: Int.MAX_VALUE }
        val patch = ordered.joinToString("\n") { jsonString(it.argumentsJson, "hunk").orEmpty() }
        return ToolCall(
            id = approvedHunks.first().id.substringBefore(":hunk-"),
            name = "apply_patch",
            argumentsJson = "{\"path\":\"${escape(path)}\",\"patch\":\"${escape(patch)}\"}",
        )
    }

    private fun jsonString(json: String, key: String): String? {
        val match = Regex("\\\"${Regex.escape(key)}\\\"\\s*:\\s*\\\"((?:\\\\.|[^\\\"\\\\])*)\\\"").find(json)
            ?: return null
        return Regex("\\\\(?:[\\\"\\\\/bfnrt]|u[0-9a-fA-F]{4})").replace(match.groupValues[1]) { escaped ->
            when (val token = escaped.value.drop(1)) {
                "\"" -> "\""; "\\" -> "\\"; "/" -> "/"; "b" -> "\b"; "f" -> "\u000c"
                "n" -> "\n"; "r" -> "\r"; "t" -> "\t"
                else -> token.drop(1).toInt(16).toChar().toString()
            }
        }
    }

    private fun escape(value: String): String = buildString {
        value.forEach { character ->
            append(when (character) {
                '\\' -> "\\\\"; '\"' -> "\\\""; '\n' -> "\\n"; '\r' -> "\\r"; '\t' -> "\\t"
                else -> character
            })
        }
    }
}

File: flutter/lib/core/diff_hunk_approval.dart
import 'dart:convert';

import 'models.dart';

/// Splits a unified patch into independently reviewable hunks without
/// applying them, and rebuilds a full `apply_patch` from approved hunks.
/// Ported from Kotlin `DiffHunkApproval.kt` with `dart:convert` replacing
/// the hand-rolled JSON string parser.
abstract final class DiffHunkApproval {
  static List<ToolCall> expand(ToolCall call) {
    if (call.name != 'apply_patch') return [call];
    final args = _parseArguments(call.argumentsJson);
    final path = args?['path'];
    final patch = args?['patch'];
    if (path is! String || path.isEmpty || patch is! String || patch.isEmpty) {
      return [call];
    }
    final lines = patch.replaceAll('\r\n', '\n').replaceAll('\r', '\n').split('\n');
    final headers = <int>[
      for (var i = 0; i < lines.length; i++)
        if (lines[i].startsWith('@@ ')) i,
    ];
    if (headers.isEmpty) return [call];
    return [
      for (var index = 0; index < headers.length; index++)
        () {
          final start = headers[index];
          final end = index + 1 < headers.length ? headers[index + 1] : lines.length;
          final hunk = lines.sublist(start, end).join('\n').trimRight();
          return ToolCall(
            id: '${call.id}:hunk-${index + 1}',
            name: 'apply_patch_hunk',
            argumentsJson: jsonEncode({
              'path': path,
              'hunk_index': index + 1,
              'hunk_count': headers.length,
              'hunk': hunk,
            }),
          );
        }(),
    ];
  }

  /// Rebuilds a full apply_patch invocation from the hunks a user approved.
  static ToolCall collapse(List<ToolCall> approvedHunks) {
    if (approvedHunks.isEmpty) {
      throw ArgumentError('At least one approved hunk is required');
    }
    if (approvedHunks.any((h) => h.name != 'apply_patch_hunk')) {
      throw ArgumentError('Only apply_patch hunks can be collapsed');
    }
    String? pathOf(ToolCall call) => _parseArguments(call.argumentsJson)?['path'] as String?;
    final path = pathOf(approvedHunks.first);
    if (path == null || path.trim().isEmpty) {
      throw ArgumentError('Approved hunk is missing a path');
    }
    if (approvedHunks.any((h) => pathOf(h) != path)) {
      throw ArgumentError('Approved hunks must belong to one file');
    }
    final ordered = [...approvedHunks]..sort((a, b) {
        final aIndex =
            (_parseArguments(a.argumentsJson)?['hunk_index'] as num?)?.toInt() ??
                0x7fffffff;
        final bIndex =
            (_parseArguments(b.argumentsJson)?['hunk_index'] as num?)?.toInt() ??
                0x7fffffff;
        return aIndex.compareTo(bIndex);
      });
    final patch = ordered
        .map((h) => _parseArguments(h.argumentsJson)?['hunk'] as String? ?? '')
        .join('\n');
    return ToolCall(
      id: approvedHunks.first.id.split(':hunk-').first,
      name: 'apply_patch',
      argumentsJson: jsonEncode({'path': path, 'patch': patch}),
    );
  }

  static Map<String, dynamic>? _parseArguments(String argumentsJson) {
    try {
      final decoded = jsonDecode(argumentsJson);
      return decoded is Map<String, dynamic> ? decoded : null;
    } on FormatException {
      return null;
    }
  }
}
